sound/soc/sof/pm.c

```diff
@@ -435,12 +435,63 @@ EXPORT_SYMBOL(snd_sof_set_d0_state);
 
 int snd_sof_resume(struct device *dev)
 {
+	struct snd_sof_dev *sdev = dev_get_drvdata(dev);
+	struct pci_dev *pci = to_pci_dev(sdev->dev);
+	int ret;
+
+	/* check if we are resuming from D0I3 or D3 */
+	if (sdev->d0_substate == SOF_DSP_D0I3) {
+		dev_dbg(sdev->dev, "ADSP will exit from D0i3...\n");
+		ret = snd_sof_set_d0_state(sdev, SOF_DSP_D0I0);
+		if (ret < 0) {
+			dev_err(sdev->dev, "error: failed to exit from D0I3 %d\n",
+				ret);
+			return ret;
+		}
+
+		pci_restore_state(pci);
+		disable_irq_wake(pci->irq);
+
+		/* resume DMA trace, only need send ipc */
+		ret = snd_sof_init_trace_ipc(sdev);
+		if (ret < 0) {
+			/* non fatal */
+			dev_warn(sdev->dev,
+				 "warning: failed to init trace after resume %d\n",
+				 ret);
+		}
+
+		return 0;
+	}
+
 	return sof_resume(dev, false);
 }
 EXPORT_SYMBOL(snd_sof_resume);
 
 int snd_sof_suspend(struct device *dev)
 {
+	struct snd_sof_dev *sdev = dev_get_drvdata(dev);
+	struct pci_dev *pci = to_pci_dev(sdev->dev);
+	int ret;
+
+	if (sdev->s0_suspend) {
+		dev_dbg(sdev->dev, "ADSP will try to enter D0i3 for S2idle...\n");
```

Comment:
@keyonjie this comment can be removed but if you want to keep it, it should be enough to say DSP trying to enter D0i3.

Reply:
updated.